### ISSUE TITLE
Quick Actions events are never called

### ISSUE THREAD

**Issue by Solido** [flutter/flutter#13634]
`quickActions.initialize((String shortcutType)` is never called.

Item are presented under IOS but a selection always return the default main widget
and debugger is never triggered under initialize(). Simple print never display the message.
V 0.0.2

```
Running flutter doctor...
[✓] Flutter (on Mac OS X 10.13.2 17C88, locale en-FR, channel master)
    • Flutter at /Users/user/Install/flutter
    • Framework revision e2192a90fc (4 days ago), 2017-12-15 19:10:18 -0800
    • Engine revision d945976b04
    • Tools Dart version 1.25.0-dev.11.0
    • Engine Dart version 2.0.0-edge.102a3c8e728a5e92a913cace4553995f8421f3e5

[✓] iOS toolchain - develop for iOS devices (Xcode 9.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.2, Build version 9C40b
    • ios-deploy 1.9.2
    • CocoaPods version 1.3.1
```

**Comment by Hixie** [comment]
cc @cbracken

**Comment by ejabu** [comment]
> java.lang.RuntimeException: Unable to start activity ComponentInfo{com.example.hehe/io.flutter.plugins.quickactions.QuickActionsPlugin$ShortcutHandlerActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'void io.flutter.plugin.common.MethodChannel.invokeMethod(java.lang.String, java.lang.Object)' on a null object reference
> 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2706)
> 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2767)
> 	at android.app.ActivityThread.-wrap12(ActivityThread.java)
> 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1514)
> 	at android.os.Handler.dispatchMessage(Handler.java:102)
> 	at android.os.Looper.loop(Looper.java:163)
> 	at android.app.ActivityThread.main(ActivityThread.java:6221)
> 	at java.lang.reflect.Method.invoke(Native Method)
> 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:904)
> 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:794)
> Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void io.flutter.plugin.common.MethodChannel.invokeMethod(java.lang.String, java.lang.Object)' on a null object reference
> 	at io.flutter.plugins.quickactions.QuickActionsPlugin$ShortcutHandlerActivity.onCreate(QuickActionsPlugin.java:117)
> 	at android.app.Activity.performCreate(Activity.java:6875)
> 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1119)
> 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2659)
> 	... 9 more
> 
> 
> 

I am running on Xiami 5A.

**Comment by DanielSmith1239** [comment]
I found a temporary solution to the problem. Add the following code to your AppDelegate file:
```
- (void)application:(UIApplication *)application performActionForShortcutItem:(UIApplicationShortcutItem *)shortcutItem completionHandler:(void (^)(BOOL succeeded))completionHandler {
    FlutterViewController* controller = (FlutterViewController*)self.window.rootViewController;
    
    FlutterMethodChannel* channel = [FlutterMethodChannel methodChannelWithName:@"plugins.flutter.io/quick_actions" binaryMessenger:controller];
    [channel invokeMethod:@"launch" arguments:shortcutItem.type];
}
```

**Comment by artrmz** [comment]
@DanielSmith1239 Thank you, it works! :)

**Comment by AppleEducate** [comment]
Thanks @DanielSmith1239  thats works for iOS! Do we need to add anything for Android?

**Comment by AppleEducate** [comment]
@workerbee22 if you add the code to the app delete it works perfectly

**Comment by workerbee22** [comment]
@AppleEducate So you mean add the code to the AppDelegate file as noted by @DanielSmith1239 ?

**Comment by workerbee22** [comment]
Added @DanielSmith1239 code to project into /ios/runner/appdelegate.m still nothing using the Pub version of the package 0.2.2. Did a flutter clean first ... nope nothing, regardless of if the app is running or not.

**Comment by workerbee22** [comment]
@AppleEducate Could you please post your exact AppDelegate.m code?

**Comment by workerbee22** [comment]
You guys are right !!! But traps for Flutter devs not accustomed to the iOS environment are:

1. 3D Touch only supported on these phones (see Apple doc for all devices). I was using a physical iPhone 6 and iPhone XR ... so an old and new devices, BOTH do NOT support 3D Touch !!!
iPhone 6s, iPhone 6s Plus, iPhone 7, iPhone 7 Plus, iPhone 8, iPhone 8 Plus, iPhone X, iPhone Xs, iPhone Xs Max
The iPhone XR, which was released in 2018, does not have 3D Touch.

2. Simulator for example using a iPhone Xs Max. To simulate a 'long press' or 3D touch, note you cannot use a mouse. In Simulator, you need to have Hardware / Touch Pressure / Use Trackpad Force ENABLED !!! So you need a MacBook Pro with a Force enabled trackpad ... luckily mine is.

Then it works including the fix in AppDelegate.m from @DanielSmith1239 . Also seems to work if the app is running in the background, or from cold start of the app.

**Comment by xyzbilal** [comment]
> I found a temporary solution to the problem. Add the following code to your AppDelegate file:
> 
> ```objc
> - (void)application:(UIApplication *)application performActionForShortcutItem:(UIApplicationShortcutItem *)shortcutItem completionHandler:(void (^)(BOOL succeeded))completionHandler {
>     FlutterViewController* controller = (FlutterViewController*)self.window.rootViewController;
>     
>     FlutterMethodChannel* channel = [FlutterMethodChannel methodChannelWithName:@"plugins.flutter.io/quick_actions" binaryMessenger:controller];
>     [channel invokeMethod:@"launch" arguments:shortcutItem.type];
> }
> ```

is this for objective c as native language for ios? I trşed to put it my project that I use swift as IOS language but could not success it run properly

**Comment by xyzbilal** [comment]
for those who use swift as a IOS language. the code below is useful to run quick actions.

```
func application(_ application: UIApplication, performActionFor shortcutItem: UIApplicationShortcutItem, completionHandler: @escaping (Bool) -> Void) {
    let controller = window.rootViewController as? FlutterViewController

    let channel = FlutterMethodChannel(name: "plugins.flutter.io/quick_actions", binaryMessenger: controller)
    channel.invokeMethod("launch", arguments: shortcutItem.type)
}
```

**Comment by Solido** [comment]
Remember to upvote this issue so the Flutter Team can give it a priority if it deserve it.

**Comment by AppleEducate** [comment]
For anyone curious this is what the AppDelagate.swift should look like with the latest swift syntax:

``` swift
import UIKit
import Flutter

@UIApplicationMain
@objc class AppDelegate: FlutterAppDelegate {
  override func application(
    _ application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?
  ) -> Bool {
    GeneratedPluginRegistrant.register(with: self)
    return super.application(application, didFinishLaunchingWithOptions: launchOptions)
  }
    
@available(iOS 9.0, *)
override func application(_ application: UIApplication, performActionFor shortcutItem: UIApplicationShortcutItem, completionHandler: @escaping (Bool) -> Void) {
    let controller = window.rootViewController as? FlutterViewController
    
    let channel = FlutterMethodChannel(name: "plugins.flutter.io/quick_actions", binaryMessenger: controller!)
    channel.invokeMethod("launch", arguments: shortcutItem.type)
    }
}

```

**Comment by dustinsterk** [comment]
Thank you for this!  Could not figure out why those actions were not being triggered.  Worked perfect.

**Comment by oliverchu** [comment]
> For anyone curious this is what the AppDelagate.swift should look like with the latest swift syntax:
> 
> ```swift
> import UIKit
> import Flutter
> 
> @UIApplicationMain
> @objc class AppDelegate: FlutterAppDelegate {
>   override func application(
>     _ application: UIApplication,
>     didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?
>   ) -> Bool {
>     GeneratedPluginRegistrant.register(with: self)
>     return super.application(application, didFinishLaunchingWithOptions: launchOptions)
>   }
>     
> @available(iOS 9.0, *)
> override func application(_ application: UIApplication, performActionFor shortcutItem: UIApplicationShortcutItem, completionHandler: @escaping (Bool) -> Void) {
>     let controller = window.rootViewController as? FlutterViewController
>     
>     let channel = FlutterMethodChannel(name: "plugins.flutter.io/quick_actions", binaryMessenger: controller!)
>     channel.invokeMethod("launch", arguments: shortcutItem.type)
>     }
> }
> ```

=== BUILD TARGET Runner OF PROJECT Runner WITH CONFIGURATION Debug ===
    /Users/oliver/Work/Rongjing/hcf/flutter-mobile/ios/Runner/AppDelegate.swift:15:8: error: overriding declaration requires an 'override' keyword
      func application(_ application: UIApplication, performActionFor shortcutItem: UIApplicationShortcutItem, completionHandler: @escaping (Bool) -> Void) {
           ^
      override 
    Flutter.FlutterAppDelegate:61:15: note: overridden declaration is here
        open func application(_ application: UIApplication, performActionFor shortcutItem: UIApplicationShortcutItem, completionHandler: @escaping (Bool) -> Void)
                  ^
    /Users/oliver/Work/Rongjing/hcf/flutter-mobile/ios/Runner/AppDelegate.swift:17:109: error: argument type 'FlutterViewController' does not conform to expected type 'FlutterBinaryMessenger'
        let channel = FlutterMethodChannel(name: "plugins.flutter.io/quick_actions", binaryMessenger: controller!)

It doesn't work .

**Comment by britannio** [comment]
The previous Swift solutions weren't compiling
The following code works for me. 
I'm on the master branch.
```
import UIKit
import Flutter

@UIApplicationMain
@objc class AppDelegate: FlutterAppDelegate {
  override func application(
    _ application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
  ) -> Bool {
    GeneratedPluginRegistrant.register(with: self)
    return super.application(application, didFinishLaunchingWithOptions: launchOptions)
  }
  
// Required for the quick_actions package
@available(iOS 9.0, *)
override func application(_ application: UIApplication, performActionFor shortcutItem: UIApplicationShortcutItem, completionHandler: @escaping (Bool) -> Void) {
    let controller : FlutterViewController = window?.rootViewController as! FlutterViewController
    
    let channel = FlutterMethodChannel(name: "plugins.flutter.io/quick_actions", binaryMessenger: controller.binaryMessenger)
    channel.invokeMethod("launch", arguments: shortcutItem.type)
    }
}
```

**Comment by filly82** [comment]
> The previous Swift solutions weren't compiling
> The following code works for me.
> I'm on the master branch.
> 
> ```
> import UIKit
> import Flutter
> 
> @UIApplicationMain
> @objc class AppDelegate: FlutterAppDelegate {
>   override func application(
>     _ application: UIApplication,
>     didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
>   ) -> Bool {
>     GeneratedPluginRegistrant.register(with: self)
>     return super.application(application, didFinishLaunchingWithOptions: launchOptions)
>   }
>   
> // Required for the quick_actions package
> @available(iOS 9.0, *)
> override func application(_ application: UIApplication, performActionFor shortcutItem: UIApplicationShortcutItem, completionHandler: @escaping (Bool) -> Void) {
>     let controller : FlutterViewController = window?.rootViewController as! FlutterViewController
>     
>     let channel = FlutterMethodChannel(name: "plugins.flutter.io/quick_actions", binaryMessenger: controller.binaryMessenger)
>     channel.invokeMethod("launch", arguments: shortcutItem.type)
>     }
> }
> ```

Works for me, thanks. But only if the App is already Running in the background. Unfortunately it does not work with a cold Start

**Comment by britannio** [comment]
> > The previous Swift solutions weren't compiling
> > The following code works for me.
> > I'm on the master branch.
> > ```
> > import UIKit
> > import Flutter
> > 
> > @UIApplicationMain
> > @objc class AppDelegate: FlutterAppDelegate {
> >   override func application(
> >     _ application: UIApplication,
> >     didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
> >   ) -> Bool {
> >     GeneratedPluginRegistrant.register(with: self)
> >     return super.application(application, didFinishLaunchingWithOptions: launchOptions)
> >   }
> >   
> > // Required for the quick_actions package
> > @available(iOS 9.0, *)
> > override func application(_ application: UIApplication, performActionFor shortcutItem: UIApplicationShortcutItem, completionHandler: @escaping (Bool) -> Void) {
> >     let controller : FlutterViewController = window?.rootViewController as! FlutterViewController
> >     
> >     let channel = FlutterMethodChannel(name: "plugins.flutter.io/quick_actions", binaryMessenger: controller.binaryMessenger)
> >     channel.invokeMethod("launch", arguments: shortcutItem.type)
> >     }
> > }
> > ```
> 
> Works for me, thanks. But only if the App is already Running in the background. Unfortunately it does not work with a cold Start

I'm facing this issue but only on my most recent app.

**Comment by Vanilo** [comment]
> Works for me, thanks. But only if the App is already Running in the background. Unfortunately it does not work with a cold Start

PerformActionFor::shortcutItem won't get called from cold start. Instead, the didFinishLaunchingWithOptions will get called from cold start. You need to check the launchOption in didFinishLaunhingWIthOptions for the shortcutItem, store to a variable and handle it in applicationDidBecomeActive. 

Here are the codes:

```
@UIApplicationMain
@objc class AppDelegate: FlutterAppDelegate {
    
    var launchedShortcutItem: UIApplicationShortcutItem?

    override func application(
    _ application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
    ) -> Bool {
        GeneratedPluginRegistrant.register(with: self)
        if let shortcutItem = launchOptions?[UIApplication.LaunchOptionsKey.shortcutItem] as? UIApplicationShortcutItem {
            launchedShortcutItem = shortcutItem
        }
        return super.application(application, didFinishLaunchingWithOptions: launchOptions)

    }

    override func applicationDidBecomeActive(_ application: UIApplication) {    
        if(launchedShortcutItem != nil){
            handleShortcutItem(shortcutItem: launchedShortcutItem!.type)
            launchedShortcutItem = nil
        }
    }

    override func application(_ application: UIApplication, performActionFor shortcutItem: UIApplicationShortcutItem, completionHandler: @escaping (Bool) -> Void) {
        handleShortcutItem(shortcutItem:shortcutItem.type)
    }

    private func handleShortcutItem(shortcutItem : String){
        let controller : FlutterViewController = window?.rootViewController as! FlutterViewController
        let channel = FlutterMethodChannel(name: "plugins.flutter.io/quick_actions", binaryMessenger: controller.binaryMessenger)
        channel.invokeMethod("launch", arguments: shortcutItem)
    }

}

```

**Comment by britannio** [comment]
> > Works for me, thanks. But only if the App is already Running in the background. Unfortunately it does not work with a cold Start
> 
> PerformActionFor::shortcutItem won't get called from cold start. Instead, the didFinishLaunchingWithOptions will get called from cold start. You need to check the launchOption in didFinishLaunhingWIthOptions for the shortcutItem, store to a variable and handle it in applicationDidBecomeActive.
> 
> Here are the codes:
> 
> ```
> @UIApplicationMain
> @objc class AppDelegate: FlutterAppDelegate {
>     
>     var launchedShortcutItem: UIApplicationShortcutItem?
> 
>     override func application(
>     _ application: UIApplication,
>     didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
>     ) -> Bool {
>         GeneratedPluginRegistrant.register(with: self)
>         if let shortcutItem = launchOptions?[UIApplication.LaunchOptionsKey.shortcutItem] as? UIApplicationShortcutItem {
>             launchedShortcutItem = shortcutItem
>         }
>         return super.application(application, didFinishLaunchingWithOptions: launchOptions)
> 
>     }
> 
>     override func applicationDidBecomeActive(_ application: UIApplication) {    
>         if(launchedShortcutItem != nil){
>             handleShortcutItem(shortcutItem: launchedShortcutItem!.type)
>             launchedShortcutItem = nil
>         }
>     }
> 
>     override func application(_ application: UIApplication, performActionFor shortcutItem: UIApplicationShortcutItem, completionHandler: @escaping (Bool) -> Void) {
>         handleShortcutItem(shortcutItem:shortcutItem.type)
>     }
> 
>     private func handleShortcutItem(shortcutItem : String){
>         let controller : FlutterViewController = window?.rootViewController as! FlutterViewController
>         let channel = FlutterMethodChannel(name: "plugins.flutter.io/quick_actions", binaryMessenger: controller.binaryMessenger)
>         channel.invokeMethod("launch", arguments: shortcutItem)
>     }
> 
> }
> ```

@Vanilo This works like a charm! Are there any technical limitations that prevent this plugin from making this process easier? Ideally, we wouldn't need to touch `AppDelegate.swift` as users.

**Comment by Vanilo** [comment]
@britannio I'm afraid this is the only way to do it as far as I know. There are only a few features that would require to make changes like this.

**Comment by filly82** [comment]
@Vanilo thanks man! worked for me as well

**Comment by britannio** [comment]
@Vanilo Does this look relevant? https://github.com/flutter/plugins/pull/2887

**Comment by mvanbeusekom** [comment]
Wondering if this is still an issue. A fix is supplied since version 0.5.0 and since then I cannot reproduce the problem anymore. 

The fix only listens to the `application: performActionFor:` method my tests seem to work fine for both a cold and a hot start of the App. I have also written a XCUITest to confirm the behaviour on both cold and hot start which also doesn't show any issue.

If you still experience the issue, would you please provide the following information:

- The version of the plugin you are using;
- The output of the `flutter doctor -v` command;
- The code in your `ios/Runner/AppDelegate.m` file.

**Comment by Solido** [comment]
Hi @mvanbeusekom 
Based on your response I close the issue
Thank you
